Alec 116 alec 122: allow to save user permission and update engine parameter

BenjaminJ opened the pull request:
https://issues.opennms.org/browse/ALEC-122
https://issues.opennms.org/browse/ALEC-116

https://confluence.internal.opennms.com/display/PRODDEV/ALEC-108+ALEC+New+GUI+to+Help+Expand+Access+and+Feedback

BenjaminJ committed:
ALEC-122

add alpha and beta to karaf-shell command

BenjaminJ committed:
ALEC-122-ALEC-116

O add rest endpoint to store data in db
o add rest endpoint to configure and store engine parameter

BenjaminJ committed:
ALEC-122-ALEC-116

O add rest endpoint to store data in db
o add rest endpoint to configure and store engine parameter

BenjaminJ committed:
ALEC-122-ALEC-116

O add rest endpoint to store data in db
o add rest endpoint to configure and store engine parameter

BenjaminJ committed:
ALEC-122-ALEC-116

O add rest endpoint to store data in db
o add rest endpoint to configure and store engine parameter

BenjaminJ committed:
ALEC-122-ALEC-116

O add rest endpoint to store data in db
o add rest endpoint to configure and store engine parameter

BenjaminJ committed:
ALEC-122-ALEC-116

O add rest endpoint to store data in db
o add rest endpoint to configure and store engine parameter

BenjaminJ committed:
ALEC-122-ALEC-116

O add rest endpoint to store data in db
o add rest endpoint to configure and store engine parameter

BenjaminJ committed:
ALEC-122-ALEC-116

O add rest endpoint to store data in db
o add rest endpoint to configure and store engine parameter

BenjaminJ committed:
ALEC-122-ALEC-116

O add rest endpoint to store data in db
o add rest endpoint to configure and store engine parameter

BenjaminJ committed:
ALEC-122-ALEC-116

O add rest endpoint to store data in db
o add rest endpoint to configure and store engine parameter

j-white commented on engine/cluster/src/main/java/org/opennms/alec/engine/cluster/ClusterEngine.java:
```diff
@@ -57,6 +61,7 @@ public ClusterEngine(MetricRegistry metrics) {
      */
     @Override
     public List<Cluster<AlarmInSpaceTime>> cluster(long timestampInMillis, Graph<CEVertex, CEEdge> g) {
+        LOG.debug("start clustering:\nDistanceMeasure: alarminspace");
```
Keep?

j-white commented on engine/dbscan/src/main/java/org/opennms/alec/engine/dbscan/DBScanEngineFactory.java:
```diff
@@ -28,16 +28,38 @@
-    private double beta = DBScanEngine.DEFAULT_BETA;
+    private static final Logger LOG = LoggerFactory.getLogger(DBScanEngineFactory.class);
+
+    private double epsilon;// = DBScanEngine.DEFAULT_EPSILON;
```
Can remove commented parts - no longer necessary

j-white commented on engine/dbscan/src/main/java/org/opennms/alec/engine/dbscan/DBScanEngineFactory.java:
```diff
@@ -46,7 +68,22 @@ public String getName() {
+                throw new RuntimeException(e);
+            }
+
+            LOG.error("Wrong distance measure configuration, we'll use default");
```
Would be good to output the value of the 'distanceMeasureFactory' specified, and the value of the default being used.

j-white commented on engine/itest/src/test/java/org/opennms/alec/engine/itest/Level2EngineComplianceTest.java:
```diff
@@ -84,6 +87,8 @@ public void setUp() {
 
     @Test
+    @Ignore("doesn't work with clusterEngine ?")
+    //TODO fix test
```
TODO

BenjaminJ replied:
I ignore the part for the cluster engine

j-white commented on engine/itest/src/test/java/org/opennms/alec/engine/itest/Level1EngineComplianceTest.java:
```diff
@@ -131,6 +134,8 @@ public void canCorrelateFiveAlarmsWithSameTimeOnSameResource() {
      */
     @Test
+    @Ignore("doesn't work with clusterEngine ?")
+//    TODO fix test
```
TODO

BenjaminJ replied:
I ignore the part for the cluster engine

j-white commented on features/ui/src/main/java/org/opennms/alec/rest/ALECRestImpl.java:
```diff
@@ -1,26 +1,119 @@
+    }
 
+    @Override
+    public Response getDB(String key) {
```
I think we should move away defining a generic key/value store here in the REST API and use concrete objects intead.

BenjaminJ replied:
I refactored the api

j-white commented on features/ui/src/main/java/org/opennms/alec/data/DataStore.java:
```diff
@@ -0,0 +1,181 @@
+
+import javax.sql.DataSource;
+
+public class DataStore {
```
We can pull this implementation from registry instead of defining our own: https://github.com/OpenNMS/opennms-integration-api/blob/v1.0.0/api/src/main/java/org/opennms/integration/api/v1/distributed/KeyValueStore.java#L39

BenjaminJ replied:
Not sure if this was exactly what you were talking about, but now datastore implements KeyValueStore

Bonrob2 commented:
Is there an associated documentation task with this?

BenjaminJ committed:
ALEC-122-ALEC-116

o remove useless debug log
o remove useless comment
o remove TODO (partially fixed unit test)
o add key value to parameter
o fixed restImpl bean

BenjaminJ committed:
ALEC-122-ALEC-116

o add BlueprintContextTest

BenjaminJ committed:
ALEC-122-ALEC-116

o rewrite rest api

BenjaminJ committed:
ALEC-122-ALEC-116

o rewrite rest api

BenjaminJ committed:
ALEC-122-ALEC-116

o rewrite rest api

BenjaminJ committed:
ALEC-122-ALEC-116

o rewrite rest api

BenjaminJ committed:
Rename 'key' to 'keyEnum' in KeyValue, add JsonPropertyOrder to KeyValue

o For convenience Response entity should return key before value

cgorantla commented on engine/dbscan/src/main/java/org/opennms/alec/engine/dbscan/DBScanEngineFactory.java:
```diff
@@ -28,16 +28,38 @@
+    private double epsilon;
+    private double alpha;
+    private double beta;
+    private String distanceMeasureFactory;
```
should this be `distanceMeasureFactoryName` ?

cgorantla commented on engine/dbscan/src/main/java/org/opennms/alec/engine/dbscan/DBScanEngineFactory.java:
```diff
@@ -46,7 +68,22 @@ public String getName() {
+
+            LOG.error("Wrong distance measure configuration {}, we'll use default {}", distanceMeasureFactory, AlarmInSpaceAndTimeDistanceMeasureFactory.class.getCanonicalName());
+        }
+        return new DBScanEngine(metrics, epsilon, alpha, beta, new AlarmInSpaceAndTimeDistanceMeasureFactory());
```
Shouldn't we throw exception if we don't have any ServiceReference for the factories?

if `AlarmInSpaceAndTimeDistanceMeasureFactory()` is the default one to use, make it member variable.

cgorantla commented on features/ui/src/main/java/org/opennms/alec/data/DataStore.java:
```diff
@@ -0,0 +1,195 @@
+
+import org.opennms.integration.api.v1.distributed.KeyValueStore;
+
+public class DataStore implements KeyValueStore<String> {
```
You don't need to implement `KeyValueStore`, you can just use it. meaning you don't need to write Upsert statements, they are already done for you.

Use `KeyValueStore` reference.

cgorantla replied:
See this PR https://github.com/OpenNMS/opennms-cortex-tss-plugin/pull/22/files on accessing `KeyValueStore`

BenjaminJ replied:
Should be better now

j-white commented on features/ui/src/main/java/org/opennms/alec/data/DataStore.java:
```diff
@@ -0,0 +1,195 @@
+
+import org.opennms.integration.api.v1.distributed.KeyValueStore;
+
+public class DataStore implements KeyValueStore<String> {
```
We should pull the `KeyValueStore` implementation from the OSGi  registry (like we're doing with the `javax.sql.DataSource` now) instead of defining our own.

j-white commented on features/ui/src/main/java/org/opennms/alec/rest/ALECRestImpl.java:
```diff
@@ -1,26 +1,171 @@
-        return Response.ok("pong!!").build();
+    @Override
+    public Response setConfiguration(KeyValue keyValue) {
+        LOG.debug("\n=============================================\n" +
```
Let's stick to single line log messages - we can remove the `===` and `\n`. This goes for other log statements as well.

j-white commented on features/ui/src/main/java/org/opennms/alec/rest/ALECRestImpl.java:
```diff
@@ -1,26 +1,171 @@
+    public Response setConfiguration(KeyValue keyValue) {
+        LOG.debug("\n=============================================\n" +
+                "Set Configuration: {}\n" +
+                "=============================================", keyValue.toString());
```
Avoid calling `.toString()`, the logger will do this automatically if necessary.

j-white commented on features/ui/src/main/java/org/opennms/alec/rest/ALECRestImpl.java:
```diff
@@ -1,26 +1,171 @@
+        LOG.debug("Get Configuration {}", keyEnum);
+        String value = (String) dataStore.get(keyEnum.toString(), ALEC_CONFIG).orElse("");
+        switch (keyEnum) {
+            case ENGINE:
```
Can we remove this, in favour of just using the `/engine/configuration` instead?

j-white commented on features/ui/src/main/java/org/opennms/alec/rest/ALECRestImpl.java:
```diff
@@ -1,26 +1,171 @@
+                } catch (JsonProcessingException e) {
+                    return somethingWentWrong(e);
+                }
+            case AGREEMENT:
```
We should follow a similar pattern that we did the engine configuration - modelling with a DTO and it's own endpoint as opposed to this generic pattern.

j-white commented on features/ui/src/test/java/org/opennms/alec/MockDataSource.java:
```diff
@@ -0,0 +1,97 @@
+
+import javax.sql.DataSource;
+
+public class MockDataSource implements DataSource {
```
It looks like this is only used to the test the Blueprint - we should use `mockito` or similar instead of defining the class.

j-white commented on pom.xml:
```diff
@@ -90,6 +90,9 @@
         <tensorflow.version>1.13.1</tensorflow.version>
+        <camel-test-blueprint.version>2.19.1</camel-test-blueprint.version>
+        <jersey-client.version>2.36</jersey-client.version>
+        <opennms.version>31.0.0-SNAPSHOT</opennms.version>
```
I think we can remove this one.

j-white reviewed (commented):
Looking better.

Happy to walk through some of this together if you have any questions on my comments.

cgorantla commented on features/ui/src/main/java/org/opennms/alec/rest/ALECRestImpl.java:
```diff
@@ -1,26 +1,171 @@
+    public Response setEngineConfiguration(EngineParameter engineParameter) {
+        try {
+            //Retrieve Driver, only one driver is registered
+            ServiceReference<?>[] engineRegistryRefs = bundleContext.getAllServiceReferences(EngineRegistry.class.getCanonicalName(), null);
```
If there is only one implementation fo EngineRegistry ( at least that's what I see currently).  Can you make it directly available with service reference ?
` <service ref="driver" interface="org.opennms.alec.engine.api.EngineRegistry"/>`

cgorantla commented on features/ui/src/main/java/org/opennms/alec/rest/ALECRestImpl.java:
```diff
@@ -1,26 +1,171 @@
+
+                //Retrieve Engines list
+                ServiceReference<?>[] engineFactoryRefs = bundleContext.getAllServiceReferences(EngineFactory.class.getCanonicalName(), null);
+                for (ServiceReference<?> engineFactoryRef : engineFactoryRefs) {
```
Possible to get EngineFactory from EngineRegistry by exposing one more method in EngineRegistry ?

BenjaminJ committed:
Update from review

o use KeyStoreValue instead of implementing it
o Use Agreement object instead of
o clean up api (SetEngineConfiguration, GetEngineConfiguration, SetAgreementConfiguration, GetEngineConfiguration, GetConfigurations)
o cleanup log with multiline

BenjaminJ committed:
make Driver directly available with service reference

geraldhumphries commented on engine/dbscan/src/main/java/org/opennms/alec/engine/dbscan/DBScanEngineFactory.java:
```diff
@@ -46,7 +68,22 @@ public String getName() {
     @Override
     public AbstractClusterEngine createEngine(MetricRegistry metrics) {
-        return new DBScanEngine(metrics, epsilon, alpha, beta);
+        if(bundleContext != null) {
```
nit: missing space

geraldhumphries commented on engine/dbscan/src/main/java/org/opennms/alec/engine/dbscan/DBScanEngine.java:
```diff
@@ -72,22 +74,26 @@ public class DBScanEngine extends AbstractClusterEngine {
+    private final DistanceMeasure distanceMeasure;
     public DBScanEngine(MetricRegistry metrics) {
-        this(metrics, DEFAULT_EPSILON, DEFAULT_ALPHA, DEFAULT_BETA);
+        this(metrics, DEFAULT_EPSILON, DEFAULT_ALPHA, DEFAULT_BETA, new AlarmInSpaceAndTimeDistanceMeasureFactory());
```
There's already a `alarmInSpaceAndTimeDistanceMeasure` bean. Could we inject that bean into DBScanEngine instead of instantiating a new instance?

geraldhumphries commented on engine/dbscan/src/main/resources/OSGI-INF/blueprint/blueprint.xml:
```diff
@@ -7,16 +7,24 @@
+            <argument value="${distanceMeasure}"/>
         </bean>
     </service>
+    <service id="alarmInSpaceAndTimeDistanceMeasure" interface="org.opennms.alec.engine.api.DistanceMeasureFactory">
```
This should be called `alarmInSpaceAndTimeDistanceMeasureFactory`. Does it need to be exposed as a service?

geraldhumphries replied:
Actually, I see what this pattern is doing now. Naming it this way and exposing it as a service are important for how `DBScanEngineFactory` works. Scanning the OSGi services each time might be necessary if it needs to dynamically load new factories/services from other bundles. Otherwise, I would suggest you populate a map of `<String, DistanceMeasureFactory>`, and pull the factories out of there.

geraldhumphries replied:
Do you still need this service now?

BenjaminJ replied:
I need it in alec/features/shell/DistanceMeasureCompleter.java and alec/features/shell/ProcessAlarm.java

geraldhumphries commented on engine/itest/src/test/java/org/opennms/alec/engine/itest/DBScanEngineOptimizationTest.java:
```diff
@@ -155,7 +156,7 @@ public EngineAsFunction() throws JAXBException, IOException {
         @Override
         public double value(double[] point) {
-            DBScanEngineFactory factory = new DBScanEngineFactory();
+            DBScanEngineFactory factory = new DBScanEngineFactory(1l,2l,3l, (BundleContext)null, "");
```
nit: spacing

geraldhumphries commented on engine/itest/src/test/java/org/opennms/alec/engine/itest/Level1EngineComplianceTest.java:
```diff
@@ -135,40 +137,42 @@ public void canGenerateDeterministicResults() throws ExecutionException, Interru
-                    .withInventoryObject(""+ i % 2, "" + i % 5);
-            for (int j = 0; j < 100; j++) {
-                builder.withEvent((i+1)*(j+1),  j % 2 == 0 ? Severity.MINOR : Severity.CLEARED);
+        if(!"cluster".equals(factory.getName())) {
```
nit: spacing

geraldhumphries commented on features/shell/src/main/java/org/opennms/alec/features/shell/DistanceMeasureCompleter.java:
```diff
@@ -0,0 +1,55 @@
+import org.apache.karaf.shell.support.completers.StringsCompleter;
+import org.opennms.alec.engine.api.DistanceMeasureFactory;
+
+@Service
```
This is annotation-based as opposed to blueprint.xml, right? If the rest of the project uses XML, it might be better to stay consistent so we avoid some technical debt, unless the plan is to slowly refactor/migrate everything to annotations.

j-white replied:
These are specific annotations for shell commands. No need to try and unify these.

geraldhumphries reviewed (commented):
Some of the comments I left disappeared while I was reviewing. Hopefully they still show up.

BenjaminJ committed:
simplify bean and service reference

o use reference-list in bean initialization
o rename engine and distanceMeasure parameter in EngineParameter.java

ALEC-116, ALEC-122

geraldhumphries commented on features/ui/src/main/java/org/opennms/alec/jackson/EngineParameterImpl.java:
```diff
@@ -0,0 +1,140 @@
+
+    @Override
+    public Double getAlpha() {
+        if(Optional.ofNullable(alpha).isPresent()) {
```
Try something like this:

```
return Optional.ofNullable(alpha)
               .orElse(DBScanEngine.DEFAULT_ALPHA);
```

geraldhumphries commented on features/shell/src/main/java/org/opennms/alec/features/shell/ProcessAlarms.java:
```diff
@@ -81,9 +85,29 @@ public class ProcessAlarms implements Action {
-        final EngineFactory engineFactory = getEngineFactory();
+        EngineFactory engineFactory = getEngineFactory();
+        final DistanceMeasureFactory distanceMeasureFactory = getDistanceMeasureFactory();
+        if("dbscan".equals(engineFactory.getName())){
```
nit: spacing

geraldhumphries commented on engine/dbscan/src/main/java/org/opennms/alec/engine/dbscan/DBScanEngineFactory.java:
```diff
@@ -46,7 +68,23 @@ public String getName() {
+                .filter(e -> distanceMeasureFactoryName.equals(e.getKey()))
+                .map(Map.Entry::getValue)
+                .findFirst();
+        if(factory.isPresent()) {
```
nit: spacing

geraldhumphries commented on features/ui/src/main/java/org/opennms/alec/rest/ALECRestImpl.java:
```diff
@@ -1,26 +1,169 @@
+            Optional<EngineFactory> factory = engineFactories.stream()
+                    .filter(engineFactory -> engineName.equals(engineFactory.getName()))
+                    .findFirst();
+            if(factory.isPresent()) {
```
nit: spacing

BenjaminJ committed:
cleanup and format code

ALEC-122, ALEC-116

geraldhumphries commented on engine/dbscan/src/main/java/org/opennms/alec/engine/dbscan/DBScanEngineFactory.java:
```diff
@@ -46,7 +68,23 @@ public String getName() {
     @Override
     public AbstractClusterEngine createEngine(MetricRegistry metrics) {
-        return new DBScanEngine(metrics, epsilon, alpha, beta);
+        Optional<DistanceMeasureFactory> factory = distanceMeasureFactoryMap.entrySet()
```
Is it possible to just do `Optional.ofNullable(distanceMeasureFactoryMap.get(distanceMeasureFactoryName))`?

BenjaminJ committed:
Simplify Map

ALEC-122, ALEC-116
